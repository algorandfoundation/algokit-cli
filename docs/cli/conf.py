# Configuration file for the Sphinx documentation builder.
#
# For the full list of built-in configuration values, see the documentation:
# https://www.sphinx-doc.org/en/master/usage/configuration.html

# -- Project information -----------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information

project = "AlgoKit"
copyright = "2023, Algorand Foundation"
author = "Algorand Foundation"

# -- General configuration ---------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration

extensions = ["sphinx_click"]

exclude_patterns = ["_build", "Thumbs.db", ".DS_Store"]
smartquotes = False
