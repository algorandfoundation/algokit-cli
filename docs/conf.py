# Configuration file for the Sphinx documentation builder.
#
# For the full list of built-in configuration values, see the documentation:
# https://www.sphinx-doc.org/en/master/usage/configuration.html
import os

# -- Project information -----------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#project-information

project = "AlgoKit"
copyright = "2023, Algorand Foundation"
author = "Algorand Foundation"

# -- General configuration ---------------------------------------------------
# https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration

extensions = ["myst_parser", "sphinx_starlight_builder", "sphinx_markdown_builder"]

exclude_patterns = ["_build", "Thumbs.db", ".DS_Store", "cli/index.rst"]

myst_heading_anchors = 4
markdown_http_base = "."
starlight_http_base = "/starlight"

# TODO: Change based on env
# myst_enable_extensions = [
#     "substitution",
# ]
# myst_substitutions = {
#     "README": "",
#     "ARC1": "",
#     "ARC32": "https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0032.md"
# }
