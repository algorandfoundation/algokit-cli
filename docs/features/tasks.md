# AlgoKit Tasks

AlgoKit Tasks are a collection of handy tasks that can be used to perform various operations on Algorand blockchain.

## Features

- [Wallet Aliasing](./tasks/wallet.md) - Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.
- [Vanity Address Generation](./tasks/vanity_address.md) - Generate vanity addresses for your Algorand accounts with the AlgoKit Vanity feature. This feature allows you to generate Algorand addresses which contains a specific keyword of your choice.
- [Transfer Assets or Algos](./tasks/transfer.md) - Transfer Algos or Assets from one account to another with the AlgoKit Transfer feature. This feature allows you to transfer Algos or Assets from one account to another on Algorand blockchain.
- [Opt-(in|out) Assets](./tasks/opt.md) - Opt-in or opt-out of Algorand Asset(s). Supports single or multiple assets.
- [Signing transactions](./tasks/sign.md) - Sign goal clerk compatible Algorand transactions.
- [Sending transactions](./tasks/send.md) - Send signed goal clerk compatible Algorand transactions.
- [NFD lookups](./tasks/nfd.md) - Perform a lookup via NFD domain or address, returning the associated address or domain respectively using the AlgoKit CLI.
- [IPFS uploads](./tasks/ipfs.md) - Upload files to IPFS.
- [Asset minting](./tasks/mint.md) - Mint new fungible or non-fungible assets on Algorand.
- [Analyze TEAL code](./tasks/analyze.md) - Analyze TEAL code using [`tealer`](https://github.com/crytic/tealer) integration for common vulnerabilities.
