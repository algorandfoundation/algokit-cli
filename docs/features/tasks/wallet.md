# AlgoKit Task Wallet

Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.

## Usage

Available commands and possible usage as follows:

```bash
$ ~ algokit task wallet
Usage: algokit task wallet [OPTIONS] COMMAND [ARGS]...

Create short aliases for your addresses and accounts on AlgoKit CLI.

Options:
-h, --help Show this message and exit.

Commands:
add Add an address or account to be stored against a named alias.
get Get an address or account stored against a named alias.
list List all addresses and accounts stored against a named alias.
remove Remove an address or account stored against a named alias.
reset Remove all aliases.
```

## Commands

### Add

This command adds an address or account to be stored against a named alias. If the `--mnemonic` flag is used, it will prompt the user for a mnemonic phrase interactively using masked input. If the `--force` flag is used, it will allow overwriting an existing alias. Maximum number of aliases that can be stored at a time is 50.

```bash
$ algokit wallet add [OPTIONS] ALIAS_NAME
```

> Please note, the command is not designed to be used in CI scope, there is no option to skip interactive masked input of the mnemonic, if you want to alias an `Account` (both private and public key) entity.

#### Options

- `--address, -a TEXT`: Specifies the address of the account. This option is required.
- `--mnemonic, -m`: If specified, it prompts the user for a mnemonic phrase interactively using masked input.
- `--force, -f`: If specified, it allows overwriting an existing alias without interactive confirmation prompt.

### Get

This command retrieves an address or account stored against a named alias.

```bash
$ algokit wallet get ALIAS
```

### List

This command lists all addresses and accounts stored against a named alias. If a record contains a `private_key` it will show a boolean flag indicating whether it exists, actual private key values are never exposed. As a user you can obtain the content of the stored aliases by navigating to your dedicated password manager (see [keyring details](https://pypi.org/project/keyring/)).

```bash
$ algokit wallet list
```

### Remove

This command removes an address or account stored against a named alias.
You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet remove ALIAS  [--force | -f]
```

### Reset

This command removes all aliases. You must confirm the prompt interactively or pass `--force` | `-f` flag to ignore the prompt.

```bash
$ algokit wallet reset [--force | -f]
```

## Keyring

AlgoKit relies on the [keyring](https://pypi.org/project/keyring/) library, which provides an easy way to interact with the operating system's password manager. This abstraction allows AlgoKit to securely manage sensitive information such as mnemonics and private keys.

When you use AlgoKit to store a mnemonic, it is never printed or exposed directly in the console. Instead, the mnemonic is converted and stored as a private key in the password manager. This ensures that your sensitive information is kept secure.

To retrieve the stored mnemonic, you will need to manually navigate to your operating system's password manager. The keyring library supports a variety of password managers across different operating systems. Here are some examples:

- On macOS, it uses the Keychain Access app.
- On Windows, it uses the Credential Manager.
- On Linux, it can use Secret Service API, KWallet, or an in-memory store depending on your setup.

> Remember, AlgoKit is designed to keep your sensitive information secure however your storage is only as secure as the device on which it is stored. Always ensure to maintain good security practices on your device, especially when dealing with mnemonics that are to be used on MainNet.

## Further Reading

For in-depth details, visit the [wallet section](../../cli/index.md#wallet) in the AlgoKit CLI reference documentation.
