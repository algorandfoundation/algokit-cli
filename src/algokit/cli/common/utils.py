# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
