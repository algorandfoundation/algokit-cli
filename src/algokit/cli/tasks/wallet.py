import json
import re

import click
from algosdk import account

from algokit.cli.tasks.utils import get_private_key_from_mnemonic, validate_address
from algokit.core.tasks.wallet import (
    WALLET_ALIASING_MAX_LIMIT,
    WalletAliasingLimitError,
    add_alias,
    get_alias,
    get_aliases,
    remove_alias,
)


def _validate_alias_name(alias_name: str) -> None:
    pattern = r"^[\w-]{1,20}$"
    if not re.match(pattern, alias_name):
        raise click.ClickException(
            "Invalid alias name. It should have at most 20 characters consisting of numbers, "
            "letters, dashes, or underscores."
        )


@click.group()
def wallet() -> None:
    """Create short aliases for your addresses and accounts on AlgoKit CLI."""


@wallet.command("add")
@click.argument("alias_name", type=click.STRING)
@click.option("--address", "-a", type=click.STRING, required=True, help="The address of the account.")
@click.option(
    "--mnemonic",
    "-m",
    "use_mnemonic",
    is_flag=True,
    help="If specified then prompt the user for a mnemonic phrase interactively using masked input.",
)
@click.option("--force", "-f", is_flag=True, help="Allow overwriting an existing alias.", type=click.BOOL)
def add(*, alias_name: str, address: str, use_mnemonic: bool, force: bool) -> None:
    """Add an address or account to be stored against a named alias (at most 50 aliases)."""

    _validate_alias_name(alias_name)
    validate_address(address)

    private_key = get_private_key_from_mnemonic() if use_mnemonic else None

    if use_mnemonic:
        derived_address = account.address_from_private_key(private_key)  # type: ignore[no-untyped-call]
        if derived_address != address:
            click.echo(
                "Warning: Address from the mnemonic doesn't match the provided address. "
                "It won't work unless the account has been rekeyed."
            )

    if get_alias(alias_name) and not force:
        response = click.prompt(
            f"Alias '{alias_name}' already exists. Overwrite?",
            type=click.Choice(["y", "n"]),
            default="n",
        )
        if response == "n":
            return

    try:
        add_alias(alias_name, address, private_key)
    except WalletAliasingLimitError as ex:
        raise click.ClickException(f"Reached the max of {WALLET_ALIASING_MAX_LIMIT} aliases.") from ex
    except Exception as ex:
        raise click.ClickException("Failed to add alias") from ex
    else:
        click.echo(f"Alias '{alias_name}' added successfully.")


@wallet.command("get")
@click.argument("alias", type=click.STRING)
def get(alias: str) -> None:
    """Get an address or account stored against a named alias."""
    alias_data = get_alias(alias)

    if not alias_data:
        raise click.ClickException(f"Alias `{alias}` does not exist.")

    click.echo(
        f"Address for alias `{alias}`: {alias_data.address}"
        f"{' (🔐 includes private key)' if alias_data.private_key else ''}"
    )


@wallet.command("list")
def list_all() -> None:
    """List all addresses and accounts stored against a named alias."""

    aliases = get_aliases()

    output = [
        {
            "alias": alias_data.alias,
            "address": alias_data.address,
            "has_private_key": bool(alias_data.private_key),
        }
        for alias_data in aliases
    ]

    content = (
        json.dumps(output, indent=2)
        if output
        else "You don't have any aliases stored yet. Create one using `algokit task wallet add`."
    )

    click.echo(content)


@wallet.command("remove")
@click.argument("alias", type=click.STRING)
@click.option("--force", "-f", is_flag=True, help="Allow removing an alias without confirmation.")
def remove(*, alias: str, force: bool) -> None:
    """Remove an address or account stored against a named alias."""

    alias_data = get_alias(alias)

    if not alias_data:
        raise click.ClickException(f"Alias `{alias}` does not exist.")

    if not force:
        response = click.prompt(
            f"🚨 This is a destructive action that will remove the `{alias_data.alias}` alias. Are you sure?",
            type=click.Choice(["y", "n"]),
            default="n",
        )

        if response == "n":
            return

    remove_alias(alias)

    click.echo(f"Alias `{alias}` removed successfully.")


@wallet.command("reset")
@click.option("--force", "-f", is_flag=True, help="Allow removing all aliases without confirmation.")
def reset(*, force: bool) -> None:
    """Remove all aliases."""

    aliases = get_aliases()

    if not aliases:
        click.echo("Warning: No aliases available to reset.")
        return

    if not force:
        response = click.prompt(
            "🚨 This is a destructive action that will clear all aliases. Are you sure?",
            type=click.Choice(["y", "n"]),
            default="n",
        )

        if response == "n":
            return

    for alias_data in aliases:
        try:
            remove_alias(alias_data.alias)
        except Exception as ex:
            raise click.ClickException(f"Failed to remove alias {alias_data.alias}") from ex

    click.echo("All aliases have been cleared.")
