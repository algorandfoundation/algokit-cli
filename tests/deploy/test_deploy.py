import json
import sys
from pathlib import Path

import pytest
from _pytest.tmpdir import TempPathFactory
from algokit.core.conf import ALGOKIT_CONFIG
from algokit.core.tasks.wallet import WALLET_ALIASES_KEYRING_USERNAME
from algosdk.account import generate_account
from algosdk.mnemonic import from_private_key
from approvaltests.namer import NamerFactory
from pytest_mock import MockerFixture

from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock
from tests.utils.which_mock import WhichMock

PYTHON_EXECUTABLE = sys.executable
# need to use an escaped python executable path in config files for windows
PYTHON_EXECUTABLE_ESCAPED = PYTHON_EXECUTABLE.replace("\\", "\\\\")
# note: spaces around the string inside print are important,
# we need to test the usage of shlex.split vs str.split, to handle
# splitting inside quotes properly
TEST_PYTHON_COMMAND = "print(' test_command_invocation ')"


@pytest.fixture(autouse=True)
def which_mock(mocker: MockerFixture) -> WhichMock:
    which_mock = WhichMock()
    mocker.patch("algokit.core.deploy.shutil.which").side_effect = which_mock.which
    return which_mock


def test_algokit_config_empty_array(tmp_path_factory: TempPathFactory) -> None:
    empty_array_config = """
[deploy]
command = []
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(empty_array_config, encoding="utf-8")
    (cwd / ".env").touch()
    result = invoke(["deploy"], cwd=cwd)

    assert result.exit_code != 0
    verify(result.output)


def test_algokit_config_invalid_syntax(tmp_path_factory: TempPathFactory) -> None:
    invalid_config = """
{"dummy": "json"}
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(invalid_config, encoding="utf-8")
    (cwd / ".env").touch()
    result = invoke(["deploy"], cwd=cwd)

    assert result.exit_code != 0
    verify(result.output)


def test_algokit_config_name_overrides(
    tmp_path_factory: TempPathFactory, proc_mock: ProcMock, which_mock: WhichMock
) -> None:
    config_with_override = """
[deploy]
command = "command_a"

[deploy.localnet]
command = "command_b"

[deploy.testnet]
command = "command_c"
    """.strip()
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_override, encoding="utf-8")
    (cwd / ".env").touch()
    (cwd / ".env.localnet").touch()
    (cwd / ".env.testnet").touch()

    resolved_cmd = which_mock.add("command_c")
    proc_mock.set_output([resolved_cmd], ["picked testnet"])

    result = invoke(["deploy", "testnet"], cwd=cwd)

    assert result.exit_code == 0
    verify(result.output)


def test_algokit_config_name_no_base(
    tmp_path_factory: TempPathFactory, proc_mock: ProcMock, which_mock: WhichMock
) -> None:
    config_with_override = """
[deploy.localnet]
command = "command_a"

[deploy.testnet]
command = "command_b"
    """.strip()
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_override, encoding="utf-8")
    (cwd / ".env.localnet").touch()
    (cwd / ".env.testnet").touch()

    cmd = which_mock.add("command_a")
    proc_mock.set_output([cmd], ["picked localnet"])

    result = invoke(["deploy", "localnet"], cwd=cwd)

    assert result.exit_code == 0
    verify(result.output)


def test_command_invocation_and_command_splitting(tmp_path: Path) -> None:
    config_data = """
[deploy]
command = ["not", "used"]
    """.strip()
    (tmp_path / ALGOKIT_CONFIG).write_text(config_data, encoding="utf-8")
    result = invoke(
        [
            "deploy",
            "--command",
            f'{PYTHON_EXECUTABLE} -c "{TEST_PYTHON_COMMAND}"',
        ],
        cwd=tmp_path,
    )
    assert result.exit_code == 0
    verify(result.output.replace(PYTHON_EXECUTABLE, "<sys.executable>"))


def test_command_splitting_from_config(tmp_path: Path) -> None:
    config_data = rf"""
[deploy]
command = "{PYTHON_EXECUTABLE_ESCAPED} -c \"{TEST_PYTHON_COMMAND}\""
    """.strip()
    (tmp_path / ALGOKIT_CONFIG).write_text(config_data, encoding="utf-8")
    result = invoke("deploy", cwd=tmp_path)
    assert result.exit_code == 0
    verify(result.output.replace(PYTHON_EXECUTABLE, "<sys.executable>"))


def test_command_without_splitting_from_config(tmp_path: Path) -> None:
    config_data = rf"""
[deploy]
command = ["{PYTHON_EXECUTABLE_ESCAPED}", "-c", "{TEST_PYTHON_COMMAND}"]
    """.strip()
    (tmp_path / ALGOKIT_CONFIG).write_text(config_data, encoding="utf-8")
    result = invoke("deploy", cwd=tmp_path)
    assert result.exit_code == 0
    verify(result.output.replace(PYTHON_EXECUTABLE, "<sys.executable>"))


@pytest.mark.usefixtures("proc_mock")
def test_command_not_found_and_no_config(tmp_path: Path) -> None:
    cmd = "gm"
    result = invoke(["deploy", "--command", cmd], cwd=tmp_path)
    assert result.exit_code != 0
    verify(result.output)


def test_command_not_executable(proc_mock: ProcMock, tmp_path: Path, which_mock: WhichMock) -> None:
    cmd = "gm"
    cmd_resolved = which_mock.add(cmd)
    proc_mock.should_deny_on([cmd_resolved])
    result = invoke(["deploy", "--command", cmd], cwd=tmp_path)
    assert result.exit_code != 0
    verify(result.output)


def test_command_bad_exit_code(proc_mock: ProcMock, tmp_path: Path, which_mock: WhichMock) -> None:
    cmd = "gm"
    cmd_resolved = which_mock.add(cmd)
    proc_mock.should_bad_exit_on([cmd_resolved], output=["it is not morning"])
    result = invoke(["deploy", "--command", cmd], cwd=tmp_path)
    assert result.exit_code != 0
    verify(result.output)


def test_algokit_env_name_missing(tmp_path_factory: TempPathFactory, which_mock: WhichMock) -> None:
    config_with_override = """
[deploy.localnet]
command = "command_a"
    """.strip()
    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_override, encoding="utf-8")
    (cwd / ".env").touch()

    which_mock.add("command_a")
    result = invoke(["deploy", "localnet"], cwd=cwd)

    assert result.exit_code == 1
    verify(result.output)


def test_algokit_env_and_name_correct_set(
    tmp_path_factory: TempPathFactory, proc_mock: ProcMock, monkeypatch: pytest.MonkeyPatch, which_mock: WhichMock
) -> None:
    env_config = """
ENV_A=GENERIC_ENV_A
ENV_B=GENERIC_ENV_B
ENV_C=GENERIC_ENV_C
    """.strip()

    env_name_config = """
ENV_A=LOCALNET_ENV_A
ENV_B=LOCALNET_ENV_B
    """.strip()

    monkeypatch.setenv("ENV_A", "ENVIRON_ENV_A")

    config_with_deploy_name = """
[deploy]
command = "command_a"

[deploy.localnet]
command = "command_b"
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_deploy_name, encoding="utf-8")
    (cwd / ".env").write_text(env_config, encoding="utf-8")
    (cwd / ".env.localnet").write_text(env_name_config, encoding="utf-8")

    cmd_resolved = which_mock.add("command_b")
    proc_mock.set_output([cmd_resolved], ["picked localnet"])

    result = invoke(["deploy", "localnet"], cwd=cwd)

    assert proc_mock.called[0].env
    passed_env_vars = proc_mock.called[0].env

    assert passed_env_vars["ENV_A"] == "ENVIRON_ENV_A"  # os.environ is highest loading priority
    assert passed_env_vars["ENV_B"] == "LOCALNET_ENV_B"  # then .env.{name}
    assert passed_env_vars["ENV_C"] == "GENERIC_ENV_C"  # lastly .env

    verify(result.output)


def test_algokit_deploy_only_base_deploy_config(
    tmp_path_factory: TempPathFactory, proc_mock: ProcMock, which_mock: WhichMock
) -> None:
    config_with_only_base_deploy = """
[deploy]
command = "command_a"
    """.strip()

    env_config = """
ENV_A=GENERIC_ENV_A
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_only_base_deploy, encoding="utf-8")
    (cwd / ".env").write_text(env_config, encoding="utf-8")

    cmd_resolved = which_mock.add("command_a")
    proc_mock.set_output([cmd_resolved], ["picked base deploy command"])

    result = invoke(["deploy"], cwd=cwd)

    assert result.exit_code == 0
    assert proc_mock.called[0].env
    passed_env_vars = proc_mock.called[0].env

    assert passed_env_vars["ENV_A"] == "GENERIC_ENV_A"

    verify(result.output)


def test_ci_flag_interactivity_mode_via_env(
    tmp_path_factory: TempPathFactory,
    mocker: MockerFixture,
    monkeypatch: pytest.MonkeyPatch,
    proc_mock: ProcMock,
    which_mock: WhichMock,
) -> None:
    monkeypatch.setenv("CI", "true")

    mock_prompt = mocker.patch("click.prompt")

    config_with_only_base_deploy = """
[deploy]
command = "command_a"
environment_secrets = [
    "DEPLOYER_MNEMONIC"
]
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_only_base_deploy, encoding="utf-8")
    (cwd / ".env").touch()

    cmd_resolved = which_mock.add("command_a")
    proc_mock.set_output([cmd_resolved], ["picked base deploy command"])

    result = invoke(["deploy"], cwd=cwd)

    mock_prompt.assert_not_called()
    assert result.exit_code != 0

    verify(result.output)


def test_ci_flag_interactivity_mode_via_cli(
    tmp_path_factory: TempPathFactory,
    mocker: MockerFixture,
    proc_mock: ProcMock,
    which_mock: WhichMock,
) -> None:
    mock_prompt = mocker.patch("click.prompt")

    config_with_only_base_deploy = """
[deploy]
command = "command_a"
environment_secrets = [
    "DEPLOYER_MNEMONIC"
]
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_only_base_deploy, encoding="utf-8")
    (cwd / ".env").touch()

    cmd_resolved = which_mock.add("command_a")
    proc_mock.set_output([cmd_resolved], ["picked base deploy command"])

    result = invoke(["deploy", "--ci"], cwd=cwd)

    mock_prompt.assert_not_called()
    assert result.exit_code != 0

    verify(result.output)


# environment_secrets set
def test_secrets_prompting_via_stdin(
    tmp_path_factory: TempPathFactory,
    mocker: MockerFixture,
    proc_mock: ProcMock,
    monkeypatch: pytest.MonkeyPatch,
    which_mock: WhichMock,
) -> None:
    # ensure Github Actions CI env var is not overriding behavior
    monkeypatch.delenv("CI", raising=False)

    # mock click.prompt
    mock_prompt = mocker.patch("click.prompt", return_value="secret_value")
    config_with_only_base_deploy = """
[deploy]
command = "command_a"
environment_secrets = [
    "DEPLOYER_MNEMONIC"
]
    """.strip()

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_only_base_deploy, encoding="utf-8")
    (cwd / ".env").touch()
    cmd_resolved = which_mock.add("command_a")
    proc_mock.set_output([cmd_resolved], ["picked base deploy command"])

    result = invoke(["deploy"], cwd=cwd)
    mock_prompt.assert_called_once()  # ensure called
    assert result.exit_code == 0  # ensure success

    # assert that entered value is passed to proc run
    assert proc_mock.called[0].env
    called_env = proc_mock.called[0].env
    assert "DEPLOYER_MNEMONIC" in called_env
    assert called_env["DEPLOYER_MNEMONIC"] == "secret_value"

    verify(result.output)


def test_deploy_custom_project_dir(
    tmp_path_factory: TempPathFactory, proc_mock: ProcMock, which_mock: WhichMock
) -> None:
    cwd = tmp_path_factory.mktemp("cwd")
    custom_folder = cwd / "custom_folder"

    custom_folder.mkdir()
    (custom_folder / ALGOKIT_CONFIG).write_text(
        """
[deploy]
command = "command_a"
    """.strip(),
        encoding="utf-8",
    )
    (custom_folder / ".env.testnet").touch()
    cmd_resolved = which_mock.add("command_a")
    proc_mock.set_output([cmd_resolved], ["picked base deploy command"])

    input_answers = ["N"]

    # Below is needed for escaping the backslash in the path on Windows
    # Works on Linux as well since \\ doesn't exist in the path in such cases
    path = str(custom_folder.absolute()).replace("\\", r"\\")
    result = invoke(f"deploy testnet --path={path}", cwd=cwd, input="\n".join(input_answers))

    assert result.exit_code == 0
    verify(result.output)


def test_deploy_shutil_command_not_found(tmp_path_factory: TempPathFactory) -> None:
    cwd = tmp_path_factory.mktemp("cwd")

    (cwd / ALGOKIT_CONFIG).write_text(
        """
[deploy]
command = "command_a"
    """.strip(),
        encoding="utf-8",
    )
    (cwd / ".env").touch()

    result = invoke("deploy", cwd=cwd)

    assert result.exit_code == 1
    verify(result.output)


@pytest.mark.parametrize(
    ("alias", "env_var_name"),
    [
        ("deployer", "DEPLOYER_MNEMONIC"),
        ("dispenser", "DISPENSER_MNEMONIC"),
    ],
)
def test_deploy_dispenser_alias(
    alias: str,
    env_var_name: str,
    tmp_path_factory: TempPathFactory,
    proc_mock: ProcMock,
    monkeypatch: pytest.MonkeyPatch,
    mock_keyring: dict[str, str],
    which_mock: WhichMock,
) -> None:
    env_config = f"""
{env_var_name}=GENERIC_ENV_A
    """.strip()

    monkeypatch.setenv(env_var_name, "GENERIC_ENV_A")

    config_with_deploy_name = f"""
[deploy]
command = "command_a"
environment_secrets = [
    "{env_var_name}"
]
    """.strip()

    dummy_account_pk, dummy_account_addr = generate_account()  # type: ignore[no-untyped-call]
    mock_keyring[alias] = json.dumps({"alias": alias, "address": dummy_account_addr, "private_key": dummy_account_pk})
    mock_keyring[WALLET_ALIASES_KEYRING_USERNAME] = json.dumps([alias])

    cwd = tmp_path_factory.mktemp("cwd")
    (cwd / ALGOKIT_CONFIG).write_text(config_with_deploy_name, encoding="utf-8")
    (cwd / ".env").write_text(env_config, encoding="utf-8")
    which_mock.add("command_a")
    result = invoke(["deploy", f"--{alias}", alias], cwd=cwd)

    assert proc_mock.called[0].env
    passed_env_vars = proc_mock.called[0].env

    assert passed_env_vars[env_var_name] == from_private_key(dummy_account_pk)  # type: ignore[no-untyped-call]

    verify(result.output, options=NamerFactory.with_parameters(alias))
