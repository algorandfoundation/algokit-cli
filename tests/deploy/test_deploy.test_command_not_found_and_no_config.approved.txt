DEBUG: Deploying from project directory: {current_working_directory}
DEBUG: Loading deploy command from project config
DEBUG: Attempting to load project config from {current_working_directory}/.algokit.toml
DEBUG: No .algokit.toml file found in the project directory.
Error: Failed to resolve deploy command, 'gm' wasn't found
